### DISCUSSION TITLE
Gracefully close UI on shutdown

### DISCUSSION THREAD

**Discussion by falkoschindler** [zauberzeug/nicegui#297] [Ideas / Feature Requests]
As @smojef mentioned in discussion https://github.com/zauberzeug/nicegui/discussions/290 it would be nice to be able to gracefully shutdown NiceGUI. There is already an `app.shutdown()` to shutdown the server (requires `ui.run(..., reload=False)`). But the browser tab will show a warning about a lost connection and will try to reconnect. We should somehow tell the client about the shutdown. So the browser can either simply stop reconnecting, or show some message, or close the tab (probably not possible).

**Comment by smojef** [comment]
The most important is to stop the browser from reconnecting (which I did not find how).
Also the tab should be kept open.
A message could be shown saying "Disconnected", although we should be able to override its appearance (i.e. translation or custom ui element) or avoid it if desired.

**Comment by rodja** [comment]
Maybe we could send some last UI update to the server before actually shutting down. I think a context aware shutdown could provide this quite nicely:

```python
with app.shutdown():
    with ui.dialog(), ui.card():
         ui.label('Bye').classes('text-5xl')
```

On entering the context the shutdown command would execute the UI and on leaving it would execute a yet to be implemented "stopReconnecting" javascript function. After that the server can be stopped.

**Reply by smojef** [reply]
Seems good to me!

**Reply by falkoschindler** [reply]
But it should be `app.shutdown`, right? I guess we don't want/need client-specific shutdowns.

**Reply by rodja** [reply]
Oh, sure. I fixed the typo above.

**Comment by rodja** [comment]
Is it really a good idea to implement this with a context? Maybe it's more like a page:

```py
@app.after_shutdown()
def bye_bye():
    with ui.dialog(), ui.card():
         ui.label('Bye').classes('text-5xl')
```
